WooCommerce/Classes/Analytics/AppStartupWaitingTimeTracker.swift

```diff
@@ -0,0 +1,111 @@
+        .syncDashboardStats,
+        .syncTopPerformers,
+        .fetchExperimentAssignments,
+        .syncJITMs,
+        .syncPaymentConnectionTokens
+    ]
+
+    /// NotificationCenter
+    ///
+    private let notificationCenter: NotificationCenter
+
+    /// Stores Manager
+    ///
+    private let stores: StoresManager
+
+    init(notificationCenter: NotificationCenter = .default,
+         stores: StoresManager = ServiceLocator.stores,
+         analyticsService: Analytics = ServiceLocator.analytics,
+         currentTimeInMillis: @escaping () -> TimeInterval = { Date().timeIntervalSince1970 }) {
+        self.notificationCenter = notificationCenter
+        self.stores = stores
+
+        super.init(trackScenario: .appStartup, analyticsService: analyticsService, currentTimeInMillis: currentTimeInMillis)
+
+        startListeningToNotifications()
+    }
+
+    /// Triggers a notification for the start or end of the provided startup action.
+    ///
+    static func notify(action: NSNotification.Name, withStatus status: ActionStatus, notificationCenter: NotificationCenter = .default) {
+        notificationCenter.post(name: action, object: status)
+    }
+
+    /// Start listening to notifications that may occur on app startup.
+    ///
+    private func startListeningToNotifications() {
+        for startupAction in startupActionsToObserve {
+            notificationCenter.addObserver(self, selector: #selector(observeNotification), name: startupAction, object: nil)
+        }
+    }
+
+    /// Handle the notifications as they are observed.
+    ///
+    @objc private func observeNotification(for notification: Notification) {
+        guard let status = (notification.object as? ActionStatus) ?? (notification.object as? String).flatMap(ActionStatus.init(rawValue:)) else {
+            return
+        }
+
+        switch status {
+        case .started:
+            start(notification.name)
+        case .completed:
+            end(notification.name)
+        }
+    }
+
+    /// Start tracking the provided startup action.
+    ///
+    private func start(_ startupAction: Notification.Name) {
+        startupActionsPending.append(startupAction)
+    }
+
+    /// End the waiting time for the provided startup action.
+    /// If all startup actions are completed, evaluate the elapsed time from the init,
+    /// and send it as an analytics event.
+    ///
+    private func end(_ startupAction: Notification.Name) {
+        startupActionsPending.removeAll { $0 == startupAction }
+
+        // End the waiting time tracker when no more actions are pending
+        if startupActionsPending.isEmpty {
+            notificationCenter.removeObserver(self) // Stop listening to any notifications
+            guard stores.isAuthenticated else { // Don't track the waiting time if the user is logged out
+                return
+            }
+            super.end() // Calculate the elapsed time and trigger analytics event
```

Comment:
Is it possible to have this scenario:

```
- Notification1-start
- Notification1-end
- Slight delay
- waiting time reported
- Notification2-start
- Notification2-end
```
In that scenario, we will get an inaccurate timer. Since we post notifications from different parts of the App, I'm not sure if we can control that that doesn't happen, can we?